File: 2024/Day 1/C++/main.cpp
#include <iostream>
#include <string>
#include <regex>
#include <vector>
#include <map>

int main(int argc, char* argv[])
{
    std::regex const regex = std::regex("^(\\d+)   (\\d+)$");
    std::smatch match;
    std::string line;
    std::vector<int> left;
    std::vector<int> right;
    while (std::getline(std::cin, line))
    {
        if (std::regex_match(line, match, regex))
        {
            if (match.size() == 3)
            {
                left.push_back(std::stoi(match[1]));
                right.push_back(std::stoi(match[2]));
            }
        }
    }

    std::sort (left.begin(), left.end());
    std::sort (right.begin(), right.end());
    int result1 = 0;
    for (size_t i = 0; i < left.size(); i++)
    {
        result1 += abs(left[i] - right[i]);
    }
    std::cout << result1 <<std::endl;
    
    std::map<int, int> count;
    for (const int& r : right)
    {
        if (count.find(r) == count.end())
        {
            count[r] = 1;
        }
        else
        {
            count[r] += 1;
        }
    }
    int result2 = 0;
    for (const int& l : left)
    {
        if (count.find(l) != count.end())
        {
            result2 += l*count[l];
        }
    }
    std::cout << result2 <<std::endl;
    return 0;
}

File: 2024/Day 1/Rust/src/main.rs
use itertools::Itertools;
use std::{collections::HashMap, convert::TryFrom};

fn main() {
    let lines = std::io::stdin().lines().collect::<Result<Vec<String>, std::io::Error>>().unwrap();
    let rows: Vec<Vec<&str>> = lines.iter().map(|x| x.split("   ").collect::<Vec<_>>()).collect();
    let mut left: Vec<i32> = Vec::new();
    let mut right: Vec<i32> = Vec::new();
    for row in rows {
        left.push(row[0].parse().unwrap());
        right.push(row[1].parse().unwrap());
    }
    left.sort();
    right.sort();
    let result1: i32 = left.iter().zip(right.iter()).map(|(l, r)| i32::abs(l-r)).sum();
    println!("{}", result1);
    let counter = right.into_iter().counts();
    let result2: i32 = left.into_iter().map(|x| x*get_count(&counter, x)).sum();
    println!("{}", result2);
}

fn get_count(counter: &HashMap<i32, usize>, key: i32) -> i32 {
    return match counter.get(&key) {
        None => 0,
        Some(count) => i32::try_from(count.clone()).unwrap(),
    };
}